agent/src/models/container/mod.rs

```diff
@@ -0,0 +1,159 @@
+use std::{sync::Arc, time::Duration};
+use tokio::{task, time::sleep};
+pub mod exec_handle;
+pub mod mock;
+use bollard::{
+    container::Config,
+    exec::{CreateExecOptions, StartExecResults},
+    image::CreateImageOptions,
+    Docker,
+};
+use exec_handle::ExecResult;
+use futures_util::TryStreamExt;
+
+use super::error::Error::{
+    self, ContainerExecDetachedError, ContainerExecError, ContainerRemoveError,
+    ContainerStartError, PullImageError,
+};
+
+#[derive(Debug, Clone)]
+pub struct Container {
+    pub id: String,
+    pub config: Config<String>,
+    docker: Arc<Docker>,
+}
+
+/// Trait for container operations
+pub trait ContainerOperations {
+    /// Start the container
+    fn start(&self) -> impl std::future::Future<Output = Result<(), Error>>;
+
+    /// Execute a command in the container
+    fn exec(
+        &self,
+        command: String,
+        workdir: Option<String>,
+    ) -> impl std::future::Future<Output = Result<ExecResult, Error>>;
+
+    /// Remove the container
+    fn remove(&self) -> impl std::future::Future<Output = Result<(), Error>> + Send;
+}
+
+impl Container {
+    pub fn new(image: String, docker: Arc<Docker>) -> Self {
+        let id = format!("{:x}", rand::random::<u128>());
+        let entrypoint = Some(vec!["/bin/sh".to_string()]);
+        let config = Config {
+            entrypoint,
+            image: Some(image),
+            attach_stdin: Some(true),
+            attach_stdout: Some(true),
+            attach_stderr: Some(true),
+            open_stdin: Some(true),
+            ..Default::default()
+        };
+        Container { id, config, docker }
+    }
+}
+
+impl ContainerOperations for Container {
+    async fn start(&self) -> Result<(), Error> {
+        // Get the image
+        self.docker
+            .create_image(
+                Some(CreateImageOptions {
+                    from_image: self
+                        .config
+                        .image
+                        .clone()
+                        .unwrap_or("debian:latest".to_string()),
```

Comment:
Do we really want to fallback on debian latest ? This image doesn't contains git and will for sure crash

```suggestion
                        .unwrap_or("bitnami/git:latest".to_string()),
```

Reply:
Or maybe lets just accept the fact that a container image is mandatory to start a container. Either by returning an error or calling `unwrap()`

Reply:
Never call `unwrap()`, please always manage return error or add alternative action ; `ok_or()` for example.

Reply:
I mean, I think that it is ok to call unwrap if before hand you made sure that the container image is defined as expected

Reply:
Sorry it happens late the `ok_or` doesn't exist my bad 😏, it's best to avoid using an unwrap, either an error is raised or the case is handled.

Reply:
The `ok_or` exist and it is perfect.

Reply:
> Sorry it happens late the `ok_or` doesn't exist my bad 😏, it's best to avoid using an unwrap, either an error is raised or the case is handled.

I agree with you my guy